### PR TITLE
bech32: Initial Version

### PR DESCRIPTION
**Pull request by matheusd**
This is a port from the github.com/btcsuite/btcutil/bech32 package.

### CODE REVIEW

**Comment by matheusd** [comment]
@dajohi If you feel this is something important, I can include those as well. Can you point me to which tests specifically you ran?

**Comment by dajohi** [comment]
https://github.com/bitcoin/bitcoin/blob/master/src/test/bech32_tests.cpp#L48-L61

File: bech32/bech32.go [hunk 1]
```diff
@@ -13,9 +13,14 @@ const charset = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
...
-// part and the data part excluding the checksum.
-func Decode(bech string) (string, []byte, error) {
+// DecodeNoLimit decodes a bech32 encoded string, returning the human-readable
+// part and the data part excluding the checksum. This function does NOT
+// validate against the BIP-173 maximum length allowed for bech32 strings and
+// is meant for use in custom applications (such as lightning network payment
+// requests), NOT on-chain addresses.
+//
+// Note that the returned data is 5-bit (base32) encoded.
+func DecodeNoLimit(bech string) (string, []byte, error) {
```
Comment (davecgh) [inline bech32/bech32.go]: The function is still failing on length > 90

File: bech32/bech32.go [hunk 2]
```diff
@@ -0,0 +1,263 @@
+// Copyright (c) 2017 The btcsuite developers
+// Use of this source code is governed by an ISC
```
Comment (davecgh) [inline bech32/bech32.go]: Need to add The Decred devs since it has the new function now.

File: bech32/error.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
+package bech32
```
Comment (davecgh) [inline bech32/error.go]: Needs license header.

**Comment by matheusd** [comment]
Fixed the copyright and documentation issues.

Also, I've added a benchmark for number of allocations of an encode/decode cycle and refactored the functions to reduce it to the minimum (2 - one for each of encode and decode results).

File: bech32/bech32.go [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+
+// DecodeNoLimit decodes a bech32 encoded string, returning the human-readable
+// part and the data part excluding the checksum. This function does NOT
+// validate against the BIP-173 maximum length allowed for bech32 strings and
+// is meant for use in custom applications (such as lightning network payment
+// requests), NOT on-chain addresses.
+//
+// Note that the returned data is 5-bit (base32) encoded.
+func DecodeNoLimit(bech string) (string, []byte, error) {
+	// The maximum allowed length for a bech32 string is 90. It must also
```
Comment (davecgh) [inline bech32/bech32.go]: Since this is in `DecodeNoLimit` it doesn't have a maximum.

File: bech32/bech32.go [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+		for i := 0; i < 5; i++ {
+			if (b>>uint(i))&1 == 1 {
+				chk ^= gen[i]
+			}
+		}
+	}
+
+	// Account for the separator (0) between high and low bits of the HRP.
+	b := chk >> 25
+	chk = (chk&0x1ffffff)<<5 ^ 0
```
Comment (davecgh) [inline bech32/bech32.go]: Since `x^0 == x`, I would suggest leaving the xor off here.

File: bech32/bech32.go [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+				chk ^= gen[i]
+			}
+		}
+	}
+
+	if checksum == nil {
+		// A nil checksum is used during encoding, so assume all bytes are zero.
+		for v := 0; v < 6; v++ {
+			b := chk >> 25
+			chk = (chk&0x1ffffff)<<5 ^ 0
```
Comment (davecgh) [inline bech32/bech32.go]: Same here.

**Pull request merged**